hopr-socks/src/main.rs

```diff
@@ -0,0 +1,203 @@
+
+    env_logger::init();
+
+    return match opt.mode {
+        RunModeOpt::Client {
+            target_host,
+            target_port,
+            auth,
+        } => spawn_socks_client(&socks_domain, target_host, target_port, auth).await,
+        RunModeOpt::Server { request_timeout, auth } => spawn_socks_server(&socks_domain, request_timeout, auth).await,
+    };
+}
+
+async fn spawn_socks_client(
+    socks_domain: &String,
+    target_host: String,
+    target_port: u16,
+    auth: AuthMode,
+) -> Result<()> {
+    let remote_domain: String = [target_host.clone(), target_port.to_string()].join(":");
+    let config = ClientConfig::default();
+
+    // Creating a SOCKS stream to the target address through the socks server
+    let mut socks = match auth {
+        AuthMode::NoAuth => Socks5Stream::connect(socks_domain.clone(), target_host, target_port, config).await?,
+        AuthMode::Password { username, password } => {
+            Socks5Stream::connect_with_password(
+                socks_domain.clone(),
+                target_host,
+                target_port,
+                username,
+                password,
+                config,
+            )
+            .await?
+        }
+    };
+
+    // Once connection is completed, can start to communicate with the server
+    http_request(&mut socks, remote_domain).await?;
+
+    Ok(())
+}
+async fn spawn_socks_server(socks_domain: &String, timeout: u64, auth: AuthMode) -> Result<()> {
+    let mut config = ServerConfig::default();
+    config.set_request_timeout(timeout);
+
+    let config = match auth {
+        AuthMode::NoAuth => {
+            warn!("No authentication has been set!");
+            config
+        }
+        AuthMode::Password { username, password } => {
+            info!("Simple auth system has been set.");
+            config.with_authentication(SimpleUserPassword { username, password })
+        }
+    };
+
+    let config = Arc::new(config);
+    let listener = TcpListener::bind(&socks_domain).await?;
+
+    info!("Listen for socks connections @ {}", &socks_domain);
+
+    // Standard TCP loop
+    loop {
+        match listener.accept().await {
+            Ok((socket, _addr)) => {
+                info!("Connection from {}", socket.peer_addr()?);
+                let socket = Socks5Socket::new(socket, config.clone());
+
+                spawn_and_log_error(socket.upgrade_to_socks5());
+            }
+            Err(err) => error!("accept error = {:?}", err),
+        }
+    }
+}
+
+fn spawn_and_log_error<F, T, A>(fut: F) -> task::JoinHandle<()>
+where
+    F: Future<Output = Result<Socks5Socket<T, A>>> + Send + 'static,
+    T: AsyncRead + AsyncWrite + Unpin,
+    A: Authentication,
+{
+    task::spawn(async move {
+        if let Err(e) = fut.await {
+            error!("{:#}", &e);
+        }
+    })
+}
+
+async fn http_request<T: AsyncRead + AsyncWrite + Unpin>(stream: &mut T, domain: String) -> Result<()> {
```

Comment:
OKay, what I suggest is to create a client that returns a pollable stream and use the http_request only in the integration test to verify the behavior. So that the client creation returns a wrapped stream and there will be a method to send and read from.

Reply:
So this would remove the ability to send data from the command line basically. I would prefer to keep this feature, as it's already there, and use the method in the test directory

Reply:
Agreed, but this might be out of scope of the original epic, so might be done later.

Reply:
But you need it for testing.

Client is a library object (unless we integrate hopr-lib), used by the test.

Reply:
Removed client part. Client will be instantiated from the testing framework.